### DISCUSSION TITLE
Building AOT with ProjectReference to source generator

### DISCUSSION THREAD

**Discussion by rudiv** [dotnet/roslyn#70900] [Q&A] [answered]
We have a test project that references our source generator:

```xml
<ProjectReference Include="..\..\src\Reaper.SourceGenerator\Reaper.SourceGenerator.csproj" ReferenceOutputAssembly="false" OutputItemType="Analyzer" />
```

This works fine until we try to compile as AOT, it is trying to compile the source generator too which obviously fails due to the netstandard2.0 target.

```
1.148 /usr/share/dotnet/sdk/8.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1207: Ahead-of-time compilation is not supported for the target framework. [/src/src/Reaper.SourceGenerator/Reaper.SourceGenerator.csproj]
------
```

The Source Generator is marked as a Roslyn Component and works in other scenarios for AOT (ie. referenced from NuGet).

What is the correct way to test this? Our generator is designed to produce code that makes the project AOT friendly, but we would obviously like the ability to test this.

The workaround we're using now is to compile the source generator first, get the nupkg and create a target that references this local package. However that seems cumbersome.

**Comment by jasonmalinowski** [comment]
This looks like a bug in that something in the AOT pipeline should have filtered out this reference but didn't. I'm asking around internally for where to route this but might be a few days since we're hitting holidays in the United States.

**Reply by jasonmalinowski** [reply, marked as answer]
This bug is now filed: https://github.com/dotnet/sdk/issues/37228

**Comment by vitek-karas** [comment]
My guess would be that you publish for AOT by passing the `PublishAOT=true` on the command line (so `dotnet publish /p:PublishAOT=true`).
This hit an unfortunate combination "weird behaviors" in MSBuild which leads to the error you get. If you set the `PublishAOT` property in your app's csproj the problem should go away.
Basically, the same problem is described in this bug: https://github.com/dotnet/runtime/issues/94406

One possible future solution is this: https://github.com/dotnet/msbuild/issues/9415
But there are others we're considering as well.

In any case our recommendation is to set the `PublishAOT` property in your project file because it also turns on analyzer and changes other things which will affect even build of your app so that it's easier to detect potential problems in your app when it is AOTed.
